### ISSUE TITLE
Cannot use custom utilities with @apply in VueJS SFC

### ISSUE THREAD

**Issue by shaneostonstowe** [tailwindlabs/tailwindcss#2083]
### Describe the problem:

I have create a custom utility saved in a separate file `JRNICustomUtilities.css` imported by `postcss-import`: 
**Content of file**
```
.standard-body {
    font-size: 0.875rem;
    font-weight: 400;
}
```
**tailwind.css**
```
@import "tailwindcss/base";
@import "./jrniCustomBaseStyles.css";

@import "tailwindcss/components";

@import "tailwindcss/utilities";
@import "./JRNICustomUtilities.css";
```

I attempt to use the custom utility in a VueJS SFC `StudioHeader` in the repo linked below. 
```
<style scoped lang="postcss">
.StudioHeader {
    @apply grid grid-cols-1 bg-blue-dark p-5;
    grid-auto-flow: column;
    align-items: center;
    gap: 10px;

    &__nav-item {
        @apply text-white standard-body;
    }
}
</style>
```

If I add the custom utility in the class attribute of the element directly it works fine.  However we have established a guideline that a class attribute with more than two utility classes is to be moved to a defined class in the component and `@apply` used.  Doing this returns this error during local build: 

```
`@apply` cannot be used with `.standard-body` because `.standard-body` either cannot be found, or its actual definition includes a pseudo-selector like :hover, :active, etc. If you're sure that `.standard-body` exists, make sure that any `@import` statements are being properly processed *before* Tailwind CSS sees your CSS, as `@apply` can only be used for classes in the same CSS tree.

   7 | 
   8 |     &__nav-item {
>  9 |         @apply text-white standard-body;
     |         ^
  10 |     }
  11 | }
```

I would expect to be able to custom utilities in @apply. 
### Link to a minimal reproduction:
https://github.com/jrni-studio/vue-queuing/tree/7

**Comment by adamwathan** [comment]
See here for an explanation of the problem and the solution (it's no longer experimental now though and just works by default):

https://github.com/tailwindlabs/tailwindcss/pull/516

TL;DR, any class you want to `@apply` in a Vue component must be created through a plugin, otherwise Tailwind has no idea the class exists because all build tools process Vue style blocks in complete isolation, so Tailwind can't actually "see" the classes in any other files. Registering those classes through a plugin wires them directly into Tailwind's "brain" so it can make them "applyable" anywhere.

**Comment by shaneostonstowe** [comment]
That's great.  Thanks so much. I really enjoy working with Tailwind. Thank you so much for creating it and sharing your obvious passion.

**Comment by Kcko** [comment]
> See here for an explanation of the problem and the solution (it's no longer experimental now though and just works by default):
> 
> #516
> 
> TL;DR, any class you want to `@apply` in a Vue component must be created through a plugin, otherwise Tailwind has no idea the class exists because all build tools process Vue style blocks in complete isolation, so Tailwind can't actually "see" the classes in any other files. Registering those classes through a plugin wires them directly into Tailwind's "brain" so it can make them "applyable" anywhere.

How to register custom class with plugin? 

```
@layer components {
  .text-h1 {
    @apply font-america-m;
    @apply text-xs md:text-lg lg:text-xxl;
    // and next utility classes and rules, could be really long and complicated
}

```

**Comment by derHodrig** [comment]
@Kcko what adamwathan means is this
https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css

tailwind.config.js
```
const plugin = require('tailwindcss/plugin')

module.exports = {
  // ...
  plugins: [
    plugin(function ({ addComponents, theme }) {
      addComponents({
        '.card': {
          backgroundColor: theme(colors.white),
          borderRadius: theme(borderRadius.lg),
          padding: theme(spacing.6),
          boxShadow: theme(boxShadow.xl),
        }
      })
    })
  ]
}
``` 

I'm not satisfied with this solution

**Comment by Kcko** [comment]
> @Kcko what adamwathan means is this https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css
> 
> tailwind.config.js
> 
> ```
> const plugin = require('tailwindcss/plugin')
> 
> module.exports = {
>   // ...
>   plugins: [
>     plugin(function ({ addComponents, theme }) {
>       addComponents({
>         '.card': {
>           backgroundColor: theme(colors.white),
>           borderRadius: theme(borderRadius.lg),
>           padding: theme(spacing.6),
>           boxShadow: theme(boxShadow.xl),
>         }
>       })
>     })
>   ]
> }
> ```
> 
> I'm not satisfied with this solution

Hi, thx for reply.

Im not satisfied too, but i understand reason for that. 
Its totally isolated. Maybe in future it could works (in next version of Vue)

Do you know that you can use @apply? 
See my question for this problem and im getting really cool answer on Discord VUE channel.

const plugin = require('tailwindcss/plugin')

```
module.exports = plugin(({ addComponents }) => {
  addComponents({
    '.text-h1': {
      '@apply text-lg md:text-2xl lg:text-3xl': {},
      '@apply text-red-400 md:text-green-500 lg:text-blue-600': {},
      '@apply leading-relaxed': {},
    },
  });
```


(https://discord.com/channels/486935104384532500/1021893921556668507)

**Comment by derHodrig** [comment]
@adamwathan @Kcko But is this also true for this
```
<template>
  <div class="t-h-full t-w-full customBG"></div>
</template>

<script>
export default {
  name: 'CustomComponent',
};
</script>

<style scoped lang="postcss">
.customBG {
  @apply t-bg-blue-800;
}
</style>
```

this throws also an error `The `t-bg-blue-800` class does not exist. If `t-bg-blue-800` is a custom class, make sure it is defined within a `@layer` directive.`


https://github.com/nuxt-modules/tailwindcss/issues/536
https://stackblitz.com/edit/nuxt-starter-6leryy?file=app.vue

**Comment by Kcko** [comment]
> 

Yes, it´s okay. Does not matter on selected lang (postcss / scss = i prefer). 
You have to move your custom class (component) or utility to TW config if you want it to work in vue SFC via `@apply`

This also has one huge advantage if you use VSC and Tailwind Intellisense, you get better autocompletitions (full with definitions not only classname ...).

In general it's a bit more complicated and I take it as a workaround but I can live with it, better than nothing.

You can also try WINDICSS ;) https://windicss.org/ (it works totally different inside and could be work as you want ... and has some features that TW has not).

**Comment by SaltedSmallFish** [comment]
@Kcko @derHodrig 
You can add css preprocessor like `additionalData: "@import path/tailwind.scss;"`.
Then your custom class can use @apply in vue SFC style block.
It works for me.

**Comment by Kcko** [comment]
@SaltedSmallFish what is additionalData? Placement for this config / variable? Some link to more information about it?

**Comment by SaltedSmallFish** [comment]
@Kcko This is the css-preprocessoroptions config of a build tool, [vite css-preprocessoroptions](https://vitejs.dev/config/shared-options.html#css-preprocessoroptions). But it's not the best solution, it will import many redundant styles in the page.

**Comment by Kcko** [comment]
> @Kcko This is the css-preprocessoroptions config of a build tool, [vite css-preprocessoroptions](https://vitejs.dev/config/shared-options.html#css-preprocessoroptions). But it's not the best solution, it will import many redundant styles in the page.

Hi, thanks, but I need something that is not so tied to a specific tool, something more universal ;-), we don't use VITE on company projects.

**Comment by AndriiGera** [comment]
Any updates regarding this? Nuxt 3.11 with Tailwind 3.4 still has this issue.

**Comment by oawebdev** [comment]
> Any updates regarding this? Nuxt 3.11 with Tailwind 3.4 still has this issue.

You may try this https://github.com/nuxt-modules/tailwindcss/issues/140#issuecomment-911997165
